keps/sig-node/2371-cri-pod-container-stats/README.md

```diff
@@ -0,0 +1,416 @@
+<!-- toc -->
+- [cAdvisor-less, CRI-full Container and Pod Stats](#cadvisor-less-cri-full-container-and-pod-stats)
+  - [Summary](#summary)
+  - [Current State](#current-state)
+    - [Current Fulfiller of Summary API Stats &amp; Future Proposal](#current-fulfiller-of-summary-api-stats--future-proposal)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+      - [CRI Implementation](#cri-implementation)
+        - [ContainerStats additions](#containerstats-additions)
+        - [PodStats CRI additions](#podstats-cri-additions)
+      - [Kubelet](#kubelet)
+      - [cAdvisor](#cadvisor)
+    - [User Stories [optional]](#user-stories-optional)
+      - [Story 1](#story-1)
+      - [Story 2](#story-2)
+    - [A note on history](#a-note-on-history)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Implementation History](#implementation-history)
+  - [Drawbacks [optional]](#drawbacks-optional)
+  - [Alternatives [optional]](#alternatives-optional)
+    - [Notes [optional]](#notes-optional)
+<!-- /toc -->
+
+# cAdvisor-less, CRI-full Container and Pod Stats
+
+## Summary
+
+Kubelet relies on the [summary API](https://github.com/kubernetes/kubernetes/blob/release-1.20/staging/src/k8s.io/kubelet/pkg/apis/stats/v1alpha1/types.go) to provide node level stats, pod level stats and container level stats. 
+
+The [CRI API](https://github.com/kubernetes/cri-api) currently does not provide enough metrics to fully supply all the fields for the summary API. As such, today some summary API metrics are fetched from cAdvisor while others originate from calling into the CRI implementation. This results in an unclear origin of metrics, duplication of work done between cAdvisor and CRI, and performance implications. 
+
+This KEP aims to enhance the CRI API with enough metrics to be able to supplement the summary API directly from CRI alleviating the need for cAdvisor to collect pod and container level stats thereby ensuring only entity (CRI runtime) will collect container and pod level metrics.
+
+
+## Current State
+
+Summary API has two interfaces:
+* [cAdvisor stats provider](https://github.com/kubernetes/kubernetes/blob/release-1.20/pkg/kubelet/stats/cadvisor_stats_provider.go)
+    * Calls cAdvisor directly to obtain node, pod, and container stats
+* [CRI stats provider](https://github.com/kubernetes/kubernetes/blob/release-1.20/pkg/kubelet/stats/cri_stats_provider.go#L54)
+    * Calls CRI implementation to provide some minimum container level stats, but still relies on cAdvisor for the majority of container level stats + pod level stats + node level stats
+
+### Current Fulfiller of Summary API Stats & Future Proposal 
+
+Below is a table describing which stats come from what source now, as well a proposal of which should come from where in the future:
```

Comment:
What are the implications for Container runtime conformance?  The summary API tests are part of node conformance currently, which would imply that container runtimes _must_ implement this to be considered conformant.
It may be worth reaching out to various container runtime maintainers to ensure they can support all of these without major architectural changes.

Reply:
CRI-O is on board and ready to support the changes, I have checked with @mikebrow to check if there would be architectural changes for containerd, as well as tagged a couple of other containerd folks to get their feedback. Are there any other runtimes you think I should check with?

Reply:
I'm more concerned about the less standard runtimes, like kata, gvisor, etc., since those don't use cgroups under the hood.

Reply:
I have @fidencio working on kata compatibility (at least with CRI-O), though I have no contacts for people working on gvisor. is there anyone you'd recommend tuning into the conversation?

Reply:
@bobbypage probably has more recent contacts on that team.